### PR TITLE
Support action on Knob failure and default value

### PR DESCRIPTION
**Pull request by djungyvr**
Whenever we setup the Knob configs we would then do:

```java
Knob.setDefaultFiles(defaultFiles);
Knob.setDefaultOnConversionFailure((throwable, value) -> { rateLimitedError(throwable); return true; });
```

@goldfrapp04 adding you as a reviewer, since you looked at #36 and I'm tagging @jjhart as author of the ticket.

### CODE REVIEW

File: src/main/com/scalyr/api/knobs/Knob.java [hunk 1]
```diff
@@ -123,6 +124,41 @@ public static void setDefaultFiles(ConfigurationFile[] files) {
...
+
+  /**
+   * Called whenever {@link #converter} fails to parse a value
+   */
+  private BiFunction<Throwable, Object, java.lang.Boolean> onConversionFailure;
+
+  /**
+   * Specify the action to take for a converter failure on a per Knob basis
+   *
+   * e.g. {@code setDefaultOnConversionFailure((throwable, value) -> { logError(throwable, value); return true; });}
```
Comment (goldfrapp04) [inline src/main/com/scalyr/api/knobs/Knob.java]: s/setDefaultOnConversionFailure /setOnConversionFailure/

Comment (djungyvr) [reply src/main/com/scalyr/api/knobs/Knob.java]: ah thank you! :)

**Comment by djungyvr** [comment]
> Hmmm, first question, why do we need a default callback && an overriding callback?

I just like providing an override whenever I provide a default. I don't feel strongly about this, so happy to just keep the static one!

File: src/main/com/scalyr/api/knobs/Knob.java [hunk 2]
```diff
@@ -139,7 +175,16 @@ public Knob(java.lang.String key, Object defaultValue, Function<Object, Object>
...
     this.defaultValue = defaultValue;
-    this.converter = converter;
+    this.converter = converter == null ? null : (obj) -> {
+      try {
+        return converter.apply(obj);
+      } catch (Throwable t) {
+        if      (onConversionFailure        != null && onConversionFailure.apply(t, obj)       ) return defaultValue;
+        else if (defaultOnConversionFailure != null && defaultOnConversionFailure.apply(t, obj)) return defaultValue;
+
+        throw t;
```
Comment (goldfrapp04) [inline src/main/com/scalyr/api/knobs/Knob.java]: If `onConversionFailure` exists but evaluates to false, then we'll still evaluate `defaultOnConversionFailure`? That doesn't feel right as you've provided the override. In that case we should immediately `throw t`, no?

Comment (djungyvr) [reply src/main/com/scalyr/api/knobs/Knob.java]: shoot, nice catch thanks!

**Pull request merged**